woocommerce/woocommerce
# Try: Validate product data with JSON schema document

### Changes proposed in this Pull Request:

This PR shows how we can use JSON Schema to introduce validation logic in a sustainable way across the client and server.  It allows for _most_ of the validation needed by core fields and third party fields without needing to create custom blocks or validate on the server.
<img width="319" alt="Screenshot 2024-05-03 at 11 32 13 AM" src="https://github.com/woocommerce/woocommerce/assets/10561050/a89d0376-cdb0-4e74-ab0f-19a5c9d76c90">
<img width="355" alt="Screenshot 2024-05-03 at 10 24 45 AM" src="https://github.com/woocommerce/woocommerce/assets/10561050/599739b5-cb9f-48db-aebc-08d379546578">
<img width="348" alt="Screenshot 2024-05-03 at 10 24 40 AM" src="https://github.com/woocommerce/woocommerce/assets/10561050/c3e689e2-2f2e-49c1-8fda-4c845e9478dd">
<img width="697" alt="Screenshot 2024-05-03 at 10 24 32 AM" src="https://github.com/woocommerce/woocommerce/assets/10561050/57480b6a-fc63-4eab-bc76-034203d68b70">



Things that still need to be followed up on:

* Performance needs work.  Lots of unnecessary rerendering
* This PR only adds validation to the regular price field and generic text field.  Other fields will need logic added to them
* Some decisions need to be made about where error message strings should live and how we address unique validation cases



### How to test the changes in this Pull Request:

1. Navigate to the new product editor
2. Type a negative number in the regular price field and blur the field, note the error message
3. Remove the number and note the required error message
4. Add a valid number, but add a higher number in the sale price field, noting the error message in the regular price field
5. Add the following snippet in a pluin and note that a valid email is required:

```php
add_filter( 'woocommerce_rest_product_schema', function( $properties ) {
	$properties[ 'custom_field_property' ] = array(
		'description'  => __( 'A custom field to demo validation', 'woocommerce' ),
		'type'         => 'string',
		'context'      => array( 'view', 'edit' ),
		'pattern'      => '^\S+@\S+\.\S+$',
		'errorMessage' => array(
			'pattern' => __(
				'Please enter a valid email address',
				'woocommerce'
			),
		),
	);

	return $properties;
} );

add_filter( 'woocommerce_block_template_register', function ( $template ) {
	if ( 'simple-product' === $template->get_id() ) {
		$basic_details = $template->get_section_by_id( 'basic-details' );

		if ( $basic_details ) {
			$basic_details->add_block(
				array(
				  'id'         => 'custom_field',
				  'blockName'  => 'woocommerce/product-text-field',
				  'order'      => 13,
				  'attributes' => array(
					'label'       => 'Email',
					'property'    => 'custom_field_property',
					'placeholder' => 'Enter a valid email',
				  ),
				)
			);
		}
	}
} );
```



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
